refactor: network state to redux

fbwoolf opened the pull request:
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/stacks-wallet-web/actions/runs/3118729905).<!-- Sticky Header Marker -->

This PR proposes moving networks state from jotai to redux, except for one derived network atom `currentNetworkAtom` which can likely also be refactored out later.

- Removed: `useUpdateNetworkState`, `useUpdateNetworkKey`, `useSetTransactionRequestAtom`, and `useSetAtomSignatureRequestToken`
- Added a `networks` slice which is persisted
- Sending request network id thru the url params from background script

cc/ @kyranjamie @fbwoolf

kyranjamie commented on src/app/store/transactions/raw.hooks.ts:
```diff
@@ -1,12 +1,32 @@
+const rawTxCache = new Map();
+
+function useRawTxState() {
+  const [txId] = useRawTxIdState();
+  const { transactionsApi } = useApi();
+  if (!txId) return;
+  const match = rawTxCache.get(txId);
+  // No need to fetch again
+  if (match) return match;
+  return transactionsApi.getRawTransactionById({ txId }).then(result => {
+    const rawTx = result.raw_tx;
+    rawTxCache.set(txId, rawTx);
+    return rawTx;
+  });
+}
+
 export function useRawDeserializedTxState() {
-  return useAtomValue(rawDeserializedTxState);
+  const rawTx = useRawTxState();
+  if (!rawTx) return;
+  return deserializeTransaction(rawTx);
```
Both need `useMemo`

kyranjamie commented on src/app/pages/add-network/add-network.tsx:
```diff
@@ -55,17 +53,13 @@ export const AddNetwork = () => {
+                url: origin,
               });
-              setNetworkKey(key);
+              networksActions.currentNetworkChanged(key);
```
Do we ever allow users to add a network and _not_ change to it by default? If not, change the network inside `newNetworkAdded`

kyranjamie commented on src/app/pages/send-tokens/components/send-form-inner.tsx:
```diff
@@ -93,8 +93,8 @@ export function SendFormInner(props: SendFormInnerProps) {
       </Suspense>
-      <RecipientField error={errors.recipient} value={values.recipient} />
-      {selectedAsset?.hasMemo && <MemoField value={values.memo} error={errors.memo} />}
+      <RecipientField error={errors.recipient} value={values.recipient || ''} />
+      {selectedAsset?.hasMemo && <MemoField value={values.memo || ''} error={errors.memo} />}
```
What's the rationale here for making all the form values optional? How did we manage to not have to do this before?

kyranjamie replied:
If there are different fields in both, it'd be great if we could separate them into two interfaces

kyranjamie commented on src/app/pages/send-tokens/components/send-tokens-confirm-drawer/send-tokens-confirm-drawer.tsx:
```diff
@@ -42,9 +42,9 @@ export function SendTokensSoftwareConfirmDrawer(props: SendTokensSoftwareConfirm
-          amount={values.amount}
-          assetId={values.assetId}
-          recipient={values.recipient}
+          amount={values.amount || ''}
+          assetId={values.assetId || ''}
+          recipient={values.recipient || ''}
```
`??` would be a more appropriate operator

kyranjamie commented on src/app/pages/send-tokens/hooks/use-selected-asset.ts:
```diff
@@ -6,7 +6,7 @@ import { ftDecimals, stacksValue } from '@app/common/stacks-utils';
 import { useSelectedAssetMetadata } from '@app/store/assets/asset.hooks';
 
-export function useSelectedAsset(assetId: string) {
+export function useSelectedAsset(assetId?: string) {
```
When arguments are made optional, the responsibility of whether or not the data is present moves from the invoker to the function itself. This isn't an improvement imo.

There's no legit case where I'd want to view details of no asset. The only reason we do this is because, in some cases, the invoking code doesn't have an `assetId` yet. That's a problem for the component using this, and not the hook itself.

fbwoolf replied:
I will just change all these back, np, it was a pretty minor part of this PR ...not really the refactor. I was just addressing an error showing for me in `transaction-request`.

kyranjamie commented on src/app/query/bns/bns.query.ts:
```diff
@@ -14,13 +14,16 @@ const bnsQueryOptions = {
 }
 
-export function useGetBnsNamesOwnedByAddress(address: string) {
+export function useGetBnsNamesOwnedByAddress(address?: string) {
```
Same here, can we not make address optional? 🤔  

I'm familiar with the trouble, which boils down to hooks not being able to be called conditionally, but the code will be in a better state if we can avoid making every hook support `undefined` input. Maybe we can jam through some ideas how to better approach this app wide.

kyranjamie commented on src/app/store/common/api-clients.hooks.ts:
```diff
@@ -1,12 +1,18 @@
+  const network = useCurrentNetworkState();
+  const config = createConfig(network.url);
+  return apiClients(config);
 }
 
+// anchored (NON-microblocks)
 export function useAnchoredApi() {
-  return useAtomValue(apiClientAnchoredState);
+  const network = useCurrentNetworkState();
+  const config = createConfig(network.url, true);
+  return apiClients(config);
```
Both need `useMemo` otherwise we'll run referential equality render loops

kyranjamie commented on src/app/store/networks/networks.hooks.ts:
```diff
@@ -0,0 +1,67 @@
+      newNetworkAdded({ chainId, id, name, url }: Network) {
+        return dispatch(networksActions.newNetworkAdded({ chainId, id, name, url }));
+      },
+      networkRemoved(id: string) {
```
"network removed" describes the result of the action we want, not the action itself. `removeNetwork` would be a better name

kyranjamie commented on src/app/store/networks/networks.slice.ts:
```diff
@@ -0,0 +1,67 @@
+
+import { DEFAULT_MAINNET_SERVER, DEFAULT_TESTNET_SERVER } from '@shared/constants';
+
+export type NetworkModes = 'mainnet' | 'testnet' | 'devnet';
```
Also, type repetition:

`type DefaultNetworkNames = keyof typeof DefaultNetworkIds;`

kyranjamie commented on src/app/store/request/request.slice.ts:
```diff
@@ -0,0 +1,19 @@
+  requestNetworkId: undefined,
+};
+
+export const requestSlice = createSlice({
+  name: 'request',
+  initialState,
+  reducers: {
+    requestNetworkActivated(state, action: PayloadAction<string>) {
+      state.requestNetworkId = action.payload;
+    },
+  },
+});
```
Wondering if there's a simpler way to do this than creating a new slice just for this override behaviour, but request is too much of a generic name imo.  `networkOverride` would make more sense for now.

kyranjamie commented on src/app/store/networks/networks.slice.ts:
```diff
@@ -0,0 +1,67 @@
+  url: DEFAULT_MAINNET_SERVER,
+} as const;
+
+export const defaultNetworks = [
+  defaultCurrentNetwork,
+  {
+    chainId: ChainID.Testnet,
+    id: DefaultNetworkIds.testnet,
+    name: 'Testnet',
+    url: DEFAULT_TESTNET_SERVER,
+  },
+  {
+    chainId: ChainID.Testnet,
+    id: DefaultNetworkIds.devnet,
+    name: 'Devnet',
+    url: 'http://localhost:3999',
+  },
+] as const;
```
Think it was tider keying the defaults by id as before (would also help spreading onto the selector too)


```ts
export const defaultNetworks: Networks = {
  mainnet: {
    url: DEFAULT_MAINNET_SERVER,
    name: 'Mainnet',
    chainId: ChainID.Mainnet,
  },
  testnet: {
    url: DEFAULT_TESTNET_SERVER,
    name: 'Testnet',
    chainId: ChainID.Testnet,
  },
  devnet: {
    url: 'http://localhost:3999',
    name: 'Devnet',
    chainId: ChainID.Testnet,
  },
} as const;
```

fbwoolf replied:
You can do either with `createEntityAdapter`, happy to change if easier for you to read.

fbwoolf replied:
I feel like leaving it as an array of Networks makes more sense to me rather than having a keyed object with a prop `id` too. `createEntityAdapter` is setup to use the id prop provided to create the keyed dictionary for you so as long as the id is included, the array makes sense here? Then when you use `networkSelectors.selectEntities` it just returns a `Dictionary<Network>`.

kyranjamie replied:
Whether we store the hard coded networks as an array or dictionary I don't really mind.

The concern is only what we ask the browser to save. Persisting data on the client introduces a whole host of new things to consider when making changes later etc. We want to persist the absolute minimum data possible. Hence the ask to persist only user-added networks, and expose the defaults via a selector.

fbwoolf replied:
I get it now! Haha. Thanks.

kyranjamie reviewed (commented):
Good work Fara. 

Let's talk through some ways we can approach the form values being undefined/general input not ready cases in our hooks in our call later.

kyranjamie commented on src/background/legacy-external-message-handler.ts:
```diff
@@ -77,7 +77,7 @@ function listenForOriginTabClose({ tabId }: ListenForOriginTabCloseArgs) {
 }
 
-async function triggerRequstWindowOpen(path: RouteUrls, urlParams: URLSearchParams) {
+async function triggerRequestWindowOpen(path: RouteUrls, urlParams: URLSearchParams) {
```
Thanks 😅 🤦🏼‍♂️ 

<img width="202" alt="image" src="https://user-images.githubusercontent.com/1618764/190108907-a9f68b0f-e623-4022-aa79-9672d09b500b.png">

kyranjamie commented on src/app/pages/transaction-request/transaction-request.tsx:
```diff
@@ -55,21 +59,25 @@ function TransactionRequestBase() {
+    const reqNetworkId = findMatchingNetworkKey(requestNetwork, networks);
+    if (!reqNetworkId) return;
+    dispatch(requestActions.requestNetworkActivated(reqNetworkId));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
See https://github.com/hirosystems/stacks-wallet-web/pull/2660

This has been bothering me for a while, we shouldn't be adding all these eslint escapes, yet sometimes we need to.

fbwoolf commented on src/app/store/networks/networks.slice.ts:
```diff
@@ -0,0 +1,67 @@
+    currentNetworkChanged(state, action: PayloadAction<EntityId>) {
+      state.currentNetworkId = action.payload;
+    },
+    newNetworkAdded(state, action: PayloadAction<Network>) {
+      networksAdapter.addOne(state, action.payload);
+    },
```
Pretty sure you wanted me to change to past tense when I refactored submitted transactions? Is there a rule here in your mind?

fbwoolf replied:
You can create an issue to rename submitted transactions bc not persisted, that is fine with me. Pretty confident past tense was recommended for the onboarding actions too, which I know we would like to remove as well. 🙃 I know redux style guide says, `we recommend trying to treat actions more as "describing events that occurred"` ...so prob why I opted for past tense. I am fine with either so I'll change to present tense here.

fbwoolf commented on src/app/store/networks/networks.selectors.ts:
```diff
@@ -0,0 +1,31 @@
+const selectCurrentNetworkId = createSelector(selectNetworksSlice, state => state.currentNetworkId);
+
+export function useNetworks() {
+  return useSelector(networksSelectors.selectEntities);
```
I realized I don't understand why we need to do this? I already upserted the default networks into state to initialize state, so just using `networkSelectors.selectEntities` is getting me all the networks, correct? I can change how I implemented it, but it was suggested as a way to initialize state when using `createEntityAdapter`.

kyranjamie replied:
Why do we need users to cache hard coded data in their storage?

fbwoolf replied:
See comment below which is also addressing changing how the network state is setup.

fbwoolf replied:
Ok, got it, I didn't understand why you were asking for this change. Thanks for feedback, I get it now. 👍

Zerepcano commented:
488

El mié., 14 sept. 2022 10:54, kyranjamie ***@***.***>
escribió:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/background/legacy-external-message-handler.ts
> <https://github.com/hirosystems/stacks-wallet-web/pull/2659#discussion_r970518414>
> :
>
> > @@ -77,7 +77,7 @@ function listenForOriginTabClose({ tabId }: ListenForOriginTabCloseArgs) {
>
>    });
>
>  }
>
>
>
> -async function triggerRequstWindowOpen(path: RouteUrls, urlParams: URLSearchParams) {
>
> +async function triggerRequestWindowOpen(path: RouteUrls, urlParams: URLSearchParams) {
>
>
> Thanks 😅 🤦🏼‍♂️
>
> [image: image]
> <https://user-images.githubusercontent.com/1618764/190108907-a9f68b0f-e623-4022-aa79-9672d09b500b.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hirosystems/stacks-wallet-web/pull/2659#pullrequestreview-1107067027>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUQMRSIVGT4HNKUYSF44UCTV6GHFDANCNFSM6AAAAAAQL2EZVI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

fbwoolf commented:
@kyranjamie I moved the req network id to the background script like we discussed ...and removed the `request` slice. All seems to be working fine so I think this is ready to review again when you are free. Thx.

kyranjamie commented on src/app/store/networks/networks.hooks.ts:
```diff
@@ -0,0 +1,79 @@
+  const params = useInitialRouteSearchParams();
+  const coreApiUrl = params.get('coreApiUrl');
+  const networkChainId = params.get('networkChainId');
+  return findMatchingNetworkKey({ coreApiUrl, networkChainId, networks });
```
Does this return new reference on every render?

fbwoolf replied:
Let me check, I can put it in useMemo.

kyranjamie commented on src/app/store/networks/networks.slice.ts:
```diff
@@ -0,0 +1,67 @@
+  url: DEFAULT_MAINNET_SERVER,
+} as const;
+
+export const defaultNetworks = {
+  [DefaultNetworkIds.mainnet]: defaultCurrentNetwork,
+  [DefaultNetworkIds.testnet]: {
+    chainId: ChainID.Testnet,
+    id: DefaultNetworkIds.testnet,
+    name: 'Testnet',
+    url: DEFAULT_TESTNET_SERVER,
+  },
+  [DefaultNetworkIds.devnet]: {
+    chainId: ChainID.Testnet,
+    id: DefaultNetworkIds.devnet,
+    name: 'Devnet',
+    url: 'http://localhost:3999',
+  },
+};
```
These could better live elsewhere now not used in reducer. I would keep in constants file imo, that's where all the hard coded value are.

kyranjamie commented on src/app/store/networks/networks.ts:
```diff
@@ -0,0 +1,5 @@
+
+import { defaultCurrentNetwork, Network } from './networks.slice';
+
+export const currentNetworkAtom = atom<Network>(defaultCurrentNetwork);
```
We can derive this atom now, rather than setting duplicate state. See wallet state file where we use `get(storeAtom)`

Something like 

```ts
export const currentNetworkAtom = atom<Network>(get => {
  const store = get(storeAtom);
  // here you can possibly even use a selector to pull current network
});
```

fbwoolf replied:
Good point, thx.

fbwoolf replied:
We can't actually derive this atom bc we aren't saving the request network to state anywhere. This is handled in our hooks so just deriving an atom from state doesn't work. I got pretty far into it before I realized this. In the test-app, the app request account header will always load the wallet network rather than the one sent by the request. I think we can definitely remove this atom with an additional refactor I can tackle next though.

kyranjamie replied:
> We can't actually derive this atom bc we aren't saving the request network to state anywhere.

I hadn't realised this, though we do actually have the params in an atom https://github.com/hirosystems/stacks-wallet-web/blob/8f1972438a9e6024300fe0c77f6a41a32e6ee06c/src/app/store/common/initial-route-search-params.ts#L3

Up to you how to handle this, though would be great to avoid duplicated "dummy" atoms, and these have proven problematic, even though intended as temporary.

fbwoolf replied:
That is true, will refactor to this ...def think it is better to avoid setting a temp atom if can.

kyranjamie commented on src/background/legacy-external-message-handler.ts:
```diff
@@ -105,7 +107,15 @@ export async function handleLegacyExternalMethodFormat(
       const { urlParams, tabId } = makeSearchParamsWithDefaults(port, [['request', payload]]);
 
-      const { id } = await triggerRequstWindowOpen(RouteUrls.TransactionRequest, urlParams);
+      const { network } = getPayloadFromToken(payload);
+
+      if (network) {
+        const developerDefinedApiUrl = getCoreApiUrl(network);
+        urlParams.set('coreApiUrl', developerDefinedApiUrl);
+        urlParams.set('networkChainId', network.chainId.toString());
+      }
```
Could we pass these in the same way we do request payload?

kyranjamie replied:
Could make a function to avoid this repetition below

```ts
function getNetworkParamsFromPayload(payload: string) {
  const { network } = getPayloadFromToken(payload);

  if (!network) return [];
  const developerDefinedApiUrl = getCoreApiUrl(network);
  return [
    ['coreApiUrl', developerDefinedApiUrl],
    ['networkChainId', network.chainId.toString()]
  ];
}

const { urlParams, tabId } = makeSearchParamsWithDefaults(port, [
  ['request', payload], 
  ...getNetworkParamsFromPayload(payload)
]);
```

fbwoolf replied:
I don't remember resolving this. I'll fix, sorry.

fbwoolf commented:
@kyranjamie hoping this is ready now, but let me know what you think. I would like to make sure it gets merged before anything else major changes on dev. There are quite a few file changes here.

kyranjamie commented on src/app/store/networks/networks.selectors.ts:
```diff
@@ -0,0 +1,38 @@
+export function useCurrentNetwork(): Network {
+  const networks = useNetworks();
+  const networkId = useCurrentNetworkId();
+  return networks[networkId] ?? defaultCurrentNetwork;
```
can you check ref equality on this, I think it's good, but worth checking

kyranjamie reviewed (approved):
lgtm besides remaining comments
